src/lib.rs

```diff
@@ -1,101 +1,405 @@
+        |capture_info, file_contents, path| {
+            let (rule, rule_listener) =
+                aggregated_queries.get_rule_and_listener(capture_info.pattern_index);
+            let mut query_match_context =
+                QueryMatchContext::new(path, file_contents, rule, &config);
+            (rule_listener.on_query_match)(capture_info.node, &mut query_match_context);
+            if let Some(violations) = query_match_context.violations.take() {
+                all_violations
+                    .lock()
+                    .unwrap()
+                    .entry(path.to_owned())
+                    .or_default()
+                    .extend(violations);
+            }
+            if let Some(fixes) = query_match_context.into_pending_fixes() {
+                assert!(config.fix);
+                files_with_fixes.append(path, file_contents, &rule.meta.name, fixes);
+            }
+        },
+    )
+    .unwrap();
+    let mut all_violations = all_violations.into_inner().unwrap();
+    if !config.fix {
+        return all_violations.into_values().flatten().collect();
+    }
+    let files_with_fixes = files_with_fixes.into_inner().unwrap();
+    if files_with_fixes.is_empty() {
+        return all_violations.into_values().flatten().collect();
+    }
+    let aggregated_results_from_files_with_fixes: HashMap<
+        PathBuf,
+        (Vec<u8>, Vec<ViolationWithContext>),
+    > = files_with_fixes
+        .into_par_iter()
+        .map(
+            |(
+                path,
+                PerFilePendingFixes {
+                    mut file_contents,
+                    pending_fixes,
+                },
+            )| {
+                let mut violations: Vec<ViolationWithContext> = Default::default();
+                run_fixing_loop(
+                    &mut violations,
+                    &mut file_contents,
+                    pending_fixes,
+                    tree_sitter_grep_args.clone(),
+                    &aggregated_queries,
+                    &path,
+                    &config,
+                );
+                (path, (file_contents, violations))
+            },
+        )
+        .collect();
+    write_files(
+        aggregated_results_from_files_with_fixes
+            .iter()
+            .map(|(path, (file_contents, _))| (&**path, &**file_contents)),
+    );
+    for (path, (_, violations)) in aggregated_results_from_files_with_fixes {
+        all_violations.insert(path, violations);
+    }
+    all_violations.into_values().flatten().collect()
+}
+
+fn run_fixing_loop(
+    violations: &mut Vec<ViolationWithContext>,
+    file_contents: &mut Vec<u8>,
+    mut pending_fixes: HashMap<RuleName, Vec<PendingFix>>,
+    tree_sitter_grep_args: tree_sitter_grep::Args,
+    aggregated_queries: &AggregatedQueries,
+    path: &Path,
+    config: &Config,
+) {
+    for _ in 0..MAX_FIX_ITERATIONS {
+        apply_fixes(file_contents, pending_fixes);
+        pending_fixes = Default::default();
```

Comment:
This might be a little weird how I pass `pending_fixes` by-value to `apply_fixes` and then allocate a fresh `HashMap`?

But so basically the fixing loop applies any fixes from the last time through the loop, clears the fixes, and then re-runs linting for the file (against the current "fixed" file contents, which don't live on disk yet) (aggregating new fixes and any other violations (which will only "survive" if this ends up being the last time through the fixing loop))